### ISSUE TITLE
OnChunkUnload callbacks on TEs get called after updating

### ISSUE THREAD

**Issue by bonii-xx** [MinecraftForge/MinecraftForge#1981]
The order in which things happen is currently: Chunk unloads -> TEs get marked for removal -> all unmarked TEs get update() calls -> all marked TEs get onChunkUnload() calls.

The issue here is that TEs, especially multiblocks, access other TEs in their update() routine, however the tile entities that are getting unloaded didn't have time to clean up their state yet and are also not invalidated yet.

The usual (expected) behaviour would be to invalidate TEs on chunk unload. However the above behaviour means that this is impossible. This has apparently been the case since forever and is probably one of the predominant reasons for chunk-related problems with TEs.

The proposed fix to this would be to move the onChunkUnload from updateEntities to markTileEntityForRemoval.
Current addition of onChunkUnload in updateEntities: https://github.com/MinecraftForge/MinecraftForge/blob/master/patches/minecraft/net/minecraft/world/World.java.patch#L420-L430

The change seems to be save, markTileEntityForRemoval is only called by the chunk on unloading. The issue has been found by skyboy and he has a custom unload function in cofhcore which seems to work fine. I've taken a look at the code in question and it looks like the only thing that could cause problems is code that (unknowingly) works around this issue. However it shouldn't be a problem for 1.8. ;)

I'm also gonna add a request for a onChunkLoad callback for TEs here. There seem to be many TEs that tick because they need to load their state after getting loaded, however they have to do so on their first update tick since there is no other place where everything else is loaded already. This would allow to convert some ticking TEs to non-ticking.

TL;DR: skyboy found the reason for about every chunk/TE fuckup and we should really fix it.

**Comment by bonii-xx** [comment]
To clarify the issue: This is not an actual bug per se, but a way to give modders a way to clean up their TileEntities before other logic runs and tries to accesses TEs that are in the unloaded chunk.

The standard solution is to have every code that accesses TEs that are not in the same block as the calling logic (or otherwise guaranteed to be in the same chunk) check for blockExists/isChunkLoaded before accessing. **This should be done anyway.**

The proposed solution with moving the unload call to markTileEntityForRemoval or adding a separate call there has the problem that that function is called from chunkloading which is done in a separate thread, which can therefore lead to race conditions.

I'm not sure if there is any easy way to add a proper improvement to help TEs aquire their proper state before everything else updates without violating the standard order of things and that is also threadsafe.

**Comment by CovertJaguar** [comment]
I am also seeing issues with tile entities not being marked invalid when a chunk unloads. This pretty much prevents the use of any kind of tile caching to reduce the number of calls to the world.

**Comment by bs2609** [comment]
This is probably resolved now by MC 1.12.2.

**Comment by williewillus** [comment]
by vanilla or a forge change?

**Comment by bs2609** [comment]
https://bugs.mojang.com/browse/MC-79154

TE removals now get processed before updates, which resolves the issue here.
